### PR TITLE
fix(reviews): auto-skip already-replied Qodo sticky findings without pushback

### LINKED ISSUE #546: bug(reviews): autoqodo loop posts duplicate consolidated comments every cycle
## Problem

The `--autoqodo` polling loop posts the same consolidated PR comment every poll cycle, flooding the PR with duplicate `@qodo-code-review` replies.

PR #470 (#469) was supposed to fix this by adding `already_replied` dedup — but the fix only sets a flag without actually preventing re-posting.

Observed on [myk-org/github-webhook-server#1133](https://github.com/myk-org/github-webhook-server/pull/1133).

**Related PRs:** #466, #470, #514, #533

## What Happens

1. Qodo posts a sticky summary comment with unresolved findings
2. autoqodo picks them up → LLM fixes code → commits → pushes → posts a consolidated PR comment (`@qodo-code-review: The following review comments were reviewed...`)
3. Findings are stored in the DB via `reviews store`
4. Next poll cycle: `reviews fetch` runs again. Qodo's sticky comment **still lists the same findings** — Qodo doesn't auto-resolve them just because we replied; it only clears findings after re-analyzing the code (or stays silent)
5. `process_and_categorize` looks up the DB → matches → sets `already_replied = True` + `previous_reply`
6. But does **not** set `is_auto_skipped` — so the finding remains "pending" and actionable
7. The review-handler prompt rules (hardened in #514) explicitly say: *"Sticky findings are ALWAYS actionable — NEVER skip. `already_replied` is CONTEXT, not a reason to skip"*
8. LLM obeys → processes the finding again → posts another identical consolidated comment
9. Repeat every cycle → PR gets spammed with duplicate comments

## Root Cause

PR #470 added `already_replied` as a context signal but deliberately chose NOT to set `is_auto_skipped`, stating:

> `already_replied` is NOT `is_auto_skipped` — auto-skip is about dismissed comments from old PRs. `already_replied` is same-PR dedup for sticky findings we already posted a reply to.

The intent was for the LLM to use `already_replied` as context and decide not to re-post. But subsequent prompt hardening (#514) explicitly told the LLM to **never** skip sticky findings regardless of `already_replied`. The code dedup signal and the prompt rules contradict each other — the LLM obeys the prompt and re-posts every cycle.

LLM-level gating is unreliable for this. The fix must be in code.

## Affected Files

- `myk_pi_tools/reviews/fetch.py` — `process_and_categorize()`, lines ~986-988
- `myk_pi_tools/reviews/fetch.py` — `_enrich_findings_with_qodo_replies()`, line ~841
- `prompts/review-handler.md` — prompt rules that forbid skipping `already_replied` findings

### Approach: Post-enrichment second pass (per rnetser's suggestion)

Keep the current pipeline order intact:
1. `process_and_categorize()` runs first — sets `already_replied = True` + `previous_reply`
2. `_enrich_findings_with_qodo_replies()` runs second — adds `qodo_response` where Qodo pushed back

Then add a **second pass after enrichment** that checks:
- `already_replied = True` AND no `qodo_response` → set `is_auto_skipped = True` + `status = "skipped"`

This is cleaner than moving enrichment before categorization because `_enrich_findings_with_qodo_replies` currently depends on `already_replied` being set (line ~848 gates on it). Restructuring the pipeline would introduce coupling. The post-enrichment pass keeps things decoupled.

### Why not move enrichment before categorization?

`_enrich_findings_with_qodo_replies` gates on `already_replied` (line ~848: `if not finding.get("already_replied"): continue`). Moving it before `process_and_categorize` would mean `already_replied` isn't set yet, breaking the enrichment logic. A second pass avoids this entirely.

### Implementation steps

1. **Add post-enrichment auto-skip pass** in `fetch.py` after `_enrich_findings_with_qodo_replies()` call (~line 1347):
   ```python
   # Post-enrichment: auto-skip already-replied findings where Qodo didn't push back
   for finding in categorized.get("qodo", []):
       if finding.get("already_replied") and not finding.get("qodo_response"):
           finding["is_auto_skipped"] = True
           finding["status"] = "skipped"
           finding["skip_reason"] = "Already replied, Qodo did not push back"
   ```

2. **Update review-handler.md** — remove the contradiction. The prompt currently says "sticky findings are ALWAYS actionable — NEVER skip" which overrides the code signal. Update to align: `is_auto_skipped` sticky findings should be presented to the user as auto-skipped (not hidden, but not actionable).

3. **Add tests** verifying both paths:
   - `already_replied=True` + no `qodo_response` → `is_auto_skipped=True`
   - `already_replied=True` + `qodo_response` present → remains actionable (no `is_auto_skipped`)

## Done

- [ ] Add post-enrichment auto-skip pass in `fetch.py` after `_enrich_findings_with_qodo_replies()`
- [ ] Auto-skip `already_replied` Qodo sticky findings when no `qodo_response` exists
- [ ] Keep findings with `qodo_response` actionable (Qodo pushed back → needs re-fix)
- [ ] Update review-handler.md prompt rules to align with code behavior
- [ ] Add test: sticky finding with `already_replied=True` + no `qodo_response` → `is_auto_skipped=True`
- [ ] Add test: sticky finding with `already_replied=True` + `qodo_response` present → remains actionable

### PR DESCRIPTION
**Pull request by myakove**
## Summary

Fixes the autoqodo polling loop posting duplicate consolidated PR comments every cycle. When a Qodo sticky finding has been replied to (`already_replied=True`) and Qodo didn't push back (no `qodo_response`), the finding is now auto-skipped instead of being re-processed.

## Root Cause

PR #470 added `already_replied` as a context signal but didn't set `is_auto_skipped`. PR #514's prompt hardening then told the LLM to "NEVER skip sticky findings" — the code signal and prompt contradicted each other, causing the LLM to re-post every cycle.

## Fix

**Post-enrichment auto-skip pass** (per @rnetser's suggestion):

1. `process_and_categorize()` runs first — sets `already_replied=True` + `previous_reply`
2. `_enrich_findings_with_qodo_replies()` runs second — adds `qodo_response` where Qodo pushed back
3. **New:** `auto_skip_replied_findings()` runs third — if `already_replied=True` AND no `qodo_response` → `is_auto_skipped=True`

Findings where Qodo DID push back (`qodo_response` exists) remain actionable.

## Changes

- `myk_pi_tools/reviews/fetch.py` — new `auto_skip_replied_findings()` function + call after enrichment
- `prompts/review-handler.md` — updated Qodo sticky rules to align with code (removed "NEVER auto-skip" contradiction)
- `tests/python/test_auto_skip_replied.py` — 10 tests covering all edge cases

## Testing

- 67 Python tests pass (including 10 new)
- 48 Node tests pass
- All tox environments green

Closes #546

### CODE REVIEW

File: tests/python/test_auto_skip_replied.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+                "qodo_response": "Still present, not fixed",
+                "status": "pending",
+                "body": "Some finding",
+            }
+        ]
+        count = auto_skip_replied_findings(findings)
+        assert count == 0
+        assert not findings[0].get("is_auto_skipped")
+        assert findings[0]["status"] == "pending"
+
```
[the remark this answers is not in the document]
Comment (myakove) [reply tests/python/test_auto_skip_replied.py:42]: Fixed — added `test_unmatched_replies_disables_auto_skip` and `test_zero_unmatched_replies_allows_auto_skip` to verify the enrichment→auto-skip pipeline behavior. The unit tests now cover the integration point between enrichment and auto-skip via the `unmatched_replies` parameter.

File: myk_pi_tools/reviews/fetch.py [hunk 2]
```diff
@@ -881,6 +881,35 @@ def _enrich_findings_with_qodo_replies(findings: list[dict[str, Any]], replies:
...
+    for finding in findings:
+        if (
+            finding.get("already_replied")
+            and not finding.get("qodo_response")
+            and not finding.get("is_auto_skipped")
+            and finding.get("status") == "pending"
+        ):
+            finding["is_auto_skipped"] = True
+            finding["status"] = "skipped"
+            finding["skip_reason"] = "Already replied, Qodo did not push back"
```
[the remark this answers is not in the document]
Comment (myakove) [reply myk_pi_tools/reviews/fetch.py:915]: Fixed — `_enrich_findings_with_qodo_replies` now marks matched replies with `_matched=True` and returns the unmatched count. `auto_skip_replied_findings` accepts `unmatched_replies` param — if any replies failed to match, auto-skip is disabled entirely to prevent missed pushback. Added test `test_unmatched_replies_disables_auto_skip`.

**Comment by myakove** [comment]
@qodo-code-review[bot]

The following review comments were reviewed and a decision was made:

### `myk_pi_tools/reviews/fetch.py:898` (qodo bug) — Pushback can be missed
> Addressed: Fixed — `_enrich_findings_with_qodo_replies` now marks matched replies with `_matched=True` and returns the unmatched count. `auto_skip_replied_findings` accepts `unmatched_replies` param — if any replies failed to match, auto-skip is disabled entirely to prevent missed pushback. Added test `test_unmatched_replies_disables_auto_skip`.
### `tests/python/test_auto_skip_replied.py:1` (qodo requirement gap) — No pipeline test for skipping
> Addressed: Fixed — added `test_unmatched_replies_disables_auto_skip` and `test_zero_unmatched_replies_allows_auto_skip` to verify the enrichment→auto-skip pipeline behavior. The unit tests now cover the integration point between enrichment and auto-skip via the `unmatched_replies` parameter.
### `tests/python/test_auto_skip_replied.py:79` (qodo finding) — test_mixed_findings tests multiple behaviors
> Addressed: Fixed — split `test_mixed_findings` into three focused tests: `test_mixed_silent_accept_is_skipped`, `test_mixed_pushback_remains_actionable`, `test_mixed_new_finding_remains_actionable`. Each test verifies exactly one behavior.

File: myk_pi_tools/reviews/fetch.py [hunk 3]
```diff
@@ -878,8 +882,54 @@ def _enrich_findings_with_qodo_replies(findings: list[dict[str, Any]], replies:
...
+    Returns the count of findings that were auto-skipped.
+    """
+    # If there are unmatched Qodo replies, don't auto-skip — pushback may be missed
+    if unmatched_replies > 0:
+        print_stderr(
+            f"Warning: {unmatched_replies} Qodo reply(s) couldn't be matched to findings"
+            " — skipping auto-skip to prevent missed pushback"
+        )
+        return 0
+
```
[the remark this answers is not in the document]
Comment (myakove) [reply myk_pi_tools/reviews/fetch.py]: Fixed — same approach as finding #2: per-finding `_enrichment_checked` flag replaces global unmatched_replies disable.

**Comment by myakove** [comment]
@qodo-code-review[bot]

The following review comments were reviewed and a decision was made:

### `myk_pi_tools/reviews/fetch.py:909` (qodo requirement gap) — unmatched_replies disables auto-skip
> Addressed: Fixed — replaced global `unmatched_replies` disable with per-finding `_enrichment_checked` flag. Each `already_replied` finding is marked `_enrichment_checked=True` during enrichment. Auto-skip only applies to findings that passed through enrichment, so stale/unmatched replies no longer block auto-skip for unrelated findings.

File: myk_pi_tools/reviews/fetch.py [hunk 4]
```diff
@@ -880,6 +884,40 @@ def _enrich_findings_with_qodo_replies(findings: list[dict[str, Any]], replies:
...
+    """
+    count = 0
+    for finding in findings:
+        if (
+            finding.get("already_replied")
+            and finding.get("_enrichment_checked")
+            and not finding.get("qodo_response")
+            and not finding.get("is_auto_skipped")
+            and finding.get("status") == "pending"
+        ):
```
[the remark this answers is not in the document]
Comment (myakove) [reply myk_pi_tools/reviews/fetch.py:912]: Fixed — removed `if qodo_replies:` guard so `_enrich_findings_with_qodo_replies()` always runs, even with empty replies. This sets `_enrichment_checked=True` on `already_replied` findings in the "Qodo is truly silent" case, enabling auto-skip. Added test `test_enrichment_with_empty_replies_sets_checked`.

**Comment by myakove** [comment]
@qodo-code-review[bot]

The following review comments were reviewed and a decision was made:

### `myk_pi_tools/reviews/fetch.py:906` (qodo bug) — Auto-skip blocked without replies
> Addressed: Fixed — removed `if qodo_replies:` guard so `_enrich_findings_with_qodo_replies()` always runs, even with empty replies. This sets `_enrichment_checked=True` on `already_replied` findings in the "Qodo is truly silent" case, enabling auto-skip. Added test `test_enrichment_with_empty_replies_sets_checked`.

**Comment by myakove** [comment]
@qodo-code-review[bot]

The following review comments were reviewed and a decision was made:

### `myk_pi_tools/reviews/fetch.py:1385` (qodo bug) — Auto-skip without reply fetch
> Addressed: Fixed — `qodo_replies` initialized to `None` (sentinel for "fetch not attempted") instead of `[]`. Enrichment guarded by `if qodo_replies is not None` — only runs when reply fetch was actually attempted. When `--user` skips the fetch, `qodo_replies` stays `None`, enrichment is skipped, `_enrichment_checked` is never set, and auto-skip correctly doesn't trigger.
